soc/arm/nordic_nrf/nrf54h/soc.c

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/cache.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <hal/nrf_hsfll.h>
+#include <hal/nrf_lrcconf.h>
+#include <soc/nrfx_coredep.h>
+LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
+
+#if defined(NRF_APPLICATION)
+#define HSFLL_NODE DT_NODELABEL(cpuapp_hsfll)
+#elif defined(NRF_RADIOCORE)
+#define HSFLL_NODE DT_NODELABEL(cpurad_hsfll)
+#else
+#error "Unsupported SoC"
+#endif
+
+#define FICR_ADDR_GET(node_id, name)                                           \
+	DT_REG_ADDR(DT_PHANDLE_BY_NAME(node_id, nordic_ficrs, name)) +         \
+		DT_PHA_BY_NAME(node_id, nordic_ficrs, name, offset)
+
+static void power_domain_init(void)
+{
+#if defined(NRF_LRCCONF010)
+	/*
+	 * Set:
+	 *  - LRCCONF010.POWERON.MAIN: 1
+	 *  - LRCCONF010.POWERON.ACT: 1
+	 *  - LRCCONF010.RETAIN.MAIN: 1
+	 *  - LRCCONF010.RETAIN.ACT: 1
+	 *
+	 *  This is done here at boot so that when the idle routine will hit
+	 *  WFI the power domain will be correctly retained.
+	 */
+
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+#endif
+
+#if defined(NRF_LRCCONF000) && defined(CONFIG_SOC_NRF54H20_ENG0_CPUAPP)
```

Comment:
question: are NRF_LRCCONF010/NRF_LRCCONF000 always defined?

Reply:
It seems they are. I'll drop these checks.